### PR TITLE
Link to bibliography from citations in subfolders

### PR DESCRIPTION
**Pull request by tchernobog**
Citations in files lower than the book source root were producing an incorrect link because they weren't considering the relative path to the root.

This commits fixes it.

### CODE REVIEW

**Comment by francisco-perez-sorrosal** [comment]
Hey @tchernobog thanks for this! By addressing this you are referring to the bug in #24 described by @mlange-42? If that's correct, I solved this too in #30 :-(

**Comment by tchernobog** [comment]
Ach, it seems indeed the same issue. I must have overlooked it; I didn't know using an absolute path would be picked up from mdBook to convert it to the proper relative URL. That solution seems therefore simpler (I still haven't tried it though).

Feel free to close this then.

**Comment by francisco-perez-sorrosal** [comment]
I'm sorry :-( But maybe we can reuse some of your tests in any case as you have some more scenarios with more nesting. If you want, I can commit #30 and you can rebase your changes and add your test additional test scenarios. In #30 I also created a test book that you can use for that too.

**Comment by francisco-perez-sorrosal** [comment]
@tchernobog I've put all my PRs in master. You can check if my solution covers your cases and update this PR in case there's something wrong by rebasing on the current master branch. If you can confirm also that your books are working ok with these changes, maybe I can the release v0.0.4 on Sunday.

**Comment by tchernobog** [comment]
@francisco-perez-sorrosal unfortunately the merged solution produces absolute paths (see screenshot).

image

Obviously this doesn't work when you want to publish online the compiled book, or move it to a separate folder.
I kindly need to ask you to reconsider my patch for merging, after reverting yours.

**Comment by tchernobog** [comment]
@francisco-perez-sorrosal I rebased my changes on top of yours.

**Comment by francisco-perez-sorrosal** [comment]
@tchernobog Thanks Matteo! I was gonna tell you that! I'll test it as soon as I can and I'll merge ASAP

File: src/lib.rs [hunk 1]
```diff
@@ -539,6 +534,29 @@ fn find_placeholders(contents: &str) -> PlaceholderIter<'_> {
     PlaceholderIter(RE.captures_iter(contents))
 }
 
+fn url_to_root(source_file: &std::path::Path) -> String {
```
Comment (francisco-perez-sorrosal) [inline src/lib.rs:537]: Maybe `get_breadcrumbs_to_root` or simply `get_dotted_path_to_root` are better names for this function

File: src/lib.rs [hunk 2]
```diff
@@ -405,10 +397,17 @@ fn replace_all_placeholders<'a>(
     let mut previous_end_index = 0;
     let mut replaced = String::new();
 
-    for placeholder in find_placeholders(s) {
+    let s = &chapter.content;
```
Comment (francisco-perez-sorrosal) [inline src/lib.rs:400]: Can you get rid of `s` and substitute it for `&chapter.content` below?

File: src/lib.rs [hunk 3]
```diff
@@ -405,10 +397,17 @@ fn replace_all_placeholders<'a>(
...
+    let chapter_path = chapter
+        .path
+        .as_ref()
+        .map(|p| p.as_path())
+        .unwrap_or_else(|| std::path::Path::new(""));
+
+    for placeholder in find_placeholders(&s) {
```
Comment (francisco-perez-sorrosal) [inline src/lib.rs:407]: ^ here

**Review by francisco-perez-sorrosal — CHANGES_REQUESTED** [review]
Looks good to me. Minor changes below. Also after addressing the changes suggested, could you please execute `cargo fmt --` to format properly the changes?

Thanks!

**Comment by tchernobog** [comment]
Done. I also added rusty-hook integration to enforce a formatting check on commit, and running tests on push.

**Comment by francisco-perez-sorrosal** [comment]
There are minor `clippy` errors in the test.

**Comment by tchernobog** [comment]
Thanks, clippy was missing the `--all-targets` flag.

**Comment by francisco-perez-sorrosal** [comment]
Yeah in my laptop was even showing different issues

**Review by francisco-perez-sorrosal — APPROVED** [review]

**Pull request merged**